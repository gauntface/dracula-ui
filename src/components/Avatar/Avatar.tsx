import React, { HTMLAttributes } from 'react'
import cx from 'classnames'
import { colors as backgroundColors } from '../Color/Box'
import { Text, TextColors } from '../Typography/Text'
import { first, last } from 'lodash'

export const AvatarVariants = {
  normal: 'drac-avatar',
  subtle: 'drac-avatar-subtle'
}

export interface AvatarProps
  extends HTMLAttributes<HTMLSpanElement & HTMLImageElement> {
  themeColor?: keyof typeof backgroundColors
  variant?: keyof typeof AvatarVariants
  src?: string
  title: string
}

export const Avatar: React.FC<AvatarProps> = props => {
  const backgroundClass = `${
    backgroundColors[props.themeColor ?? 'green']
  }-transparent`

  const classes = cx(
    'drac-avatar',
    backgroundClass,
    TextColors[props.themeColor ?? 'green'],
    AvatarVariants[props.variant ?? 'normal']
  )
  const names = props.title.split(' ')

  const f = first(names)?.slice(0, 1)
  const l = last(names)?.slice(0, 1)

  return (
    <span
      className={classes}
      {...props}
      style={{
        ...props.style,
        backgroundImage: props.src && `url('${props.src}')`
      }}
    >
      {!props.src && (
        <Text color={props.themeColor ?? 'white'}>
          {f}
          {l}
        </Text>
      )}
    </span>
  )
}
