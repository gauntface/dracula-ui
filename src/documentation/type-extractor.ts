import * as ts from 'typescript'

export function extractWithComment(
  fileNames: string[],
  options: ts.CompilerOptions
): void {
  const program = ts.createProgram(fileNames, options)
  const checker: ts.TypeChecker = program.getTypeChecker()

  for (const sourceFile of program.getSourceFiles()) {
    if (!sourceFile.isDeclarationFile) {
      ts.forEachChild(sourceFile, visit)
    }
  }

  function visit(node: ts.Node) {
    const count = node.getChildCount()

    if (count > 0) {
      ts.forEachChild(node, visit)
    }

    if (ts.isInterfaceDeclaration(node) && node.name) {
      const symbol = checker.getSymbolAtLocation(node.name)
      if (symbol) {
        return serializeSymbol(symbol)
      }
    }
  }

  function serializeSymbol(symbol: ts.Symbol) {
    const s = {
      name: symbol.getName(),
      comment: ts.displayPartsToString(symbol.getDocumentationComment(checker)),
      type: checker.typeToString(
        checker.getTypeOfSymbolAtLocation(symbol, symbol.valueDeclaration!)
      )
    }
    console.log(s)
    return s
  }
}

console.log(extractWithComment(['src/components/Avatar/Avatar.tsx'], {}))
