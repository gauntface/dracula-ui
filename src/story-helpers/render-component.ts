import testRenderer from 'react-test-renderer'

export type ComponentExample = {
  code: ReturnType<testRenderer.ReactTestRenderer['toJSON']>
  docs: string
  title: string
}

export type SnapshotBuilder = {
  comp: () => React.ReactElement
  docs: string
  title: string
}

export function buildSnapshot(
  title: string,
  comp: () => React.ReactElement,
  docs: string = ''
): SnapshotBuilder {
  return {
    title,
    comp,
    docs: docs.trim()
  }
}

export function renderSnapshot(
  title: string,
  comp: React.ReactElement,
  docs: string = ''
): ComponentExample {
  const rendered = testRenderer.create(comp)
  const code = rendered.toJSON()
  return {
    title,
    code,
    docs
  }
}
