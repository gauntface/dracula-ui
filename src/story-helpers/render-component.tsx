import React from 'react'
import Enzyme, { mount,  shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import prettier from 'prettier'
import toJson from "enzyme-to-json";

Enzyme.configure({ adapter: new Adapter() })

export type ComponentExample = {
  html: string
  docs: string
  title: string
  react?: any
}

export type SnapshotBuilder = {
  comp: () => React.ReactElement
  docs: string
  title: string
}

export function buildSnapshot(
  title: string,
  comp: () => React.ReactElement,
  docs: string = ''
): SnapshotBuilder {
  return {
    title,
    comp,
    docs: docs.trim()
  }
}

export function renderSnapshot(
  title: string,
  comp: React.ReactElement,
  docs: string = ''
): ComponentExample {

  
  const html = mount(comp).html()
  const react = toJson(shallow(<div>{comp}</div>))

  const prettyHTML = prettier.format(html, {
    parser: 'html'
  })

  return {
    title,
    html: prettyHTML,
    react,
    docs
  }
}
