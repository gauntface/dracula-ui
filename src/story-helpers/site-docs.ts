import {
  renderSnapshot,
  SnapshotBuilder
} from '@/story-helpers/render-component'
import React from 'react'

// import * as serializer from 'jest-snapshot-serializer-raw'

interface Documentation {
  basicUsage: () => SnapshotBuilder
  variations: () => Array<SnapshotBuilder>
}

export function siteDocs<T>(
  componentClass: React.FunctionComponent<T>,
  documentation: Documentation
) {
  describe(`Site: ${componentClass.displayName ?? componentClass.name}`, () => {
    test('Basic Usage', () => {
      const variation = documentation.basicUsage()
      const snapshot = renderSnapshot(
        variation.title,
        variation.comp(),
        variation.docs
      )
      expect(snapshot).toMatchSnapshot()
    })

    const variations = documentation.variations()

    variations.forEach(variation => {
      test(variation.title, () => {
        const snapshot = renderSnapshot(
          variation.title,
          variation.comp(),
          variation.docs
        )
        expect(snapshot).toMatchSnapshot()
      })
    })
  })
}
