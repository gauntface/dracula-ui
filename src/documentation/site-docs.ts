import {
  ComponentExample,
  renderSnapshot,
  SnapshotBuilder
} from '../story-helpers/render-component'
import React from 'react'
import fs from 'fs-extra'
import { pretty } from '../story-helpers/pretty'
import { getDocGen } from '../documentation/docgen/doc-generator'
import { toDSP } from '../documentation/dsp/component-generator'

interface Documentation {
  basicUsage: () => SnapshotBuilder
  variations: () => Array<SnapshotBuilder>
}

export function siteDocs<T>(
  componentClass: React.FunctionComponent<T>,
  documentation: Documentation
) {
  const name = componentClass.displayName ?? componentClass.name

  describe(`Site: ${name}`, () => {
    let examples: Record<string, ComponentExample> = {}

    afterAll(async () => {
      const docGen = getDocGen(name)
      const dsp = toDSP(name, examples, docGen)
      const dspPath = `${process.cwd()}/dsp/data/components/${name}.json`
      await fs.writeFile(dspPath, pretty(JSON.stringify(dsp), 'json'))
    })

    test('Basic Usage', () => {
      const variation = documentation.basicUsage()
      const snapshot = renderSnapshot(
        variation.title,
        variation.comp,
        variation.docs
      )
      expect(snapshot).toMatchSnapshot()
      examples = {
        ...examples,
        basic: snapshot
      }
    })

    const variations = documentation.variations()

    variations.forEach((variation) => {
      test(variation.title, () => {
        const snapshot = renderSnapshot(
          variation.title,
          variation.comp,
          variation.docs
        )
        expect(snapshot).toMatchSnapshot()

        examples = {
          ...examples,
          [variation.title]: snapshot
        }
      })
    })
  })
}
