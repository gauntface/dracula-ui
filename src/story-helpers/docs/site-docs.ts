import {
  ComponentExample,
  renderSnapshot,
  SnapshotBuilder
} from '@/story-helpers/render-component'
import React from 'react'
import fs from 'fs'
import { pretty } from '@/story-helpers/pretty'

interface Documentation {
  basicUsage: () => SnapshotBuilder
  variations: () => Array<SnapshotBuilder>
}

export function siteDocs<T>(
  componentClass: React.FunctionComponent<T>,
  documentation: Documentation
) {
  const name = componentClass.displayName ?? componentClass.name

  describe(`Site: ${name}`, () => {
    let examples: Record<string, ComponentExample> = {}

    afterAll(() => {
      const path = `${process.cwd()}/examples/${name}.json`
      fs.writeFileSync(path, pretty(JSON.stringify(examples), 'json'))
    })

    test('Basic Usage', () => {
      const variation = documentation.basicUsage()
      const snapshot = renderSnapshot(
        variation.title,
        variation.comp,
        variation.docs
      )
      expect(snapshot).toMatchSnapshot()
      examples = {
        ...examples,
        basic: snapshot
      }
    })

    const variations = documentation.variations()

    variations.forEach((variation) => {
      test(variation.title, () => {
        const snapshot = renderSnapshot(
          variation.title,
          variation.comp,
          variation.docs
        )
        expect(snapshot).toMatchSnapshot()

        examples = {
          ...examples,
          [variation.title]: snapshot
        }
      })
    })
  })
}
