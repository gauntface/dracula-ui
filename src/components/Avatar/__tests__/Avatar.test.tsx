import React from 'react'
import { docs } from '@/documentation/site-docs'
import { snapshot } from '@/documentation/render-component'
import { Avatar } from '@/components/Avatar/Avatar'

docs(Avatar, {
  basic() {
    return snapshot(
      'Usage',
      Usage,
      `The most basic instance of an Avatar requires a src image and a title to be used for accessibility or as a fallback in case the image can't be loaded.`
    )
  },
  variations() {
    return [
      snapshot(
        'Colors',
        Colors,
        `The Avatar component accepts a *color* param that allows for customizing the background and accent color for an Avatar.`
      ),
      snapshot(
        'Variants',
        Variants,
        `Avatars accept multiple variants, which can be used to render a more subtle variation of the Avatar component.`
      ),
      snapshot(
        `Borders`,
        BorderVariants,
        `A border variant allows for customization of the border width of an Avatar, or for an Avatar to be rendered without a border.`
      )
    ]
  }
})

function Usage() {
  return (
    <Avatar
      title="Count Dracula"
      src="https://ui.draculatheme.com/static/images/avatar.png"
    />
  )
}

function Colors() {
  return (
    <Avatar
      src="https://ui.draculatheme.com/static/images/avatar.png"
      title="Count Dracula"
      color="cyan"
    />
  )
}

function Variants() {
  return (
    <div>
      <Avatar
        src="https://ui.draculatheme.com/static/images/avatar.png"
        title="Count Dracula"
        variant="normal"
        mb="sm"
      />
      <Avatar
        src="https://ui.draculatheme.com/static/images/avatar.png"
        title="Count Dracula"
        variant="subtle"
      />
    </div>
  )
}

function BorderVariants() {
  return (
    <div>
      <Avatar
        src="https://ui.draculatheme.com/static/images/avatar.png"
        title="Count Dracula"
        borderVariant="large"
        mb="sm"
      />
      <Avatar
        src="https://ui.draculatheme.com/static/images/avatar.png"
        title="Count Dracula"
        borderVariant="none"
      />
    </div>
  )
}