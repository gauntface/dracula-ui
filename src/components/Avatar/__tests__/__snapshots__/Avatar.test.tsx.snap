// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Site: Avatar Basic Usage 1`] = `
Object {
  "docs": "The most basic instance of an Avatar requires a src image and a title to be used for accessibility or as a fallback in case the image can't be loaded.",
  "html": "<span class=\\"drac-avatar drac-bg-green-transparent drac-text-green\\"
  ><span
    class=\\"drac-avatar-background\\"
    style=\\"
      background-image: url(https://ui.draculatheme.com/static/images/avatar.png);
    \\"
  ></span
></span>
",
  "react": "<Avatar
  title=\\"Count Dracula\\"
  src=\\"https://ui.draculatheme.com/static/images/avatar.png\\"
/>;
",
  "title": "Usage",
}
`;

exports[`Site: Avatar Borders 1`] = `
Object {
  "docs": "A border variant allows for customization of the border width of an Avatar, or for an Avatar to be rendered without a border.",
  "html": "<div>
  <span
    class=\\"drac-avatar drac-bg-green-transparent drac-text-green drac-avatar-lg-stroke drac-mb-sm\\"
    ><span
      class=\\"drac-avatar-background\\"
      style=\\"
        background-image: url(https://ui.draculatheme.com/static/images/avatar.png);
      \\"
    ></span></span
  ><span
    class=\\"drac-avatar drac-bg-green-transparent drac-text-green drac-avatar-no-border\\"
    ><span
      class=\\"drac-avatar-background\\"
      style=\\"
        background-image: url(https://ui.draculatheme.com/static/images/avatar.png);
      \\"
    ></span
  ></span>
</div>
",
  "react": "<div>
  <Avatar
    src=\\"https://ui.draculatheme.com/static/images/avatar.png\\"
    title=\\"Count Dracula\\"
    borderVariant=\\"large\\"
    mb=\\"sm\\"
  />
  <Avatar
    src=\\"https://ui.draculatheme.com/static/images/avatar.png\\"
    title=\\"Count Dracula\\"
    borderVariant=\\"none\\"
  />
</div>;
",
  "title": "Borders",
}
`;

exports[`Site: Avatar Colors 1`] = `
Object {
  "docs": "The Avatar component accepts a *color* param that allows for customizing the background and accent color for an Avatar.",
  "html": "<span class=\\"drac-avatar drac-bg-cyan-transparent drac-text-cyan\\"
  ><span
    class=\\"drac-avatar-background\\"
    style=\\"
      background-image: url(https://ui.draculatheme.com/static/images/avatar.png);
    \\"
  ></span
></span>
",
  "react": "<Avatar
  src=\\"https://ui.draculatheme.com/static/images/avatar.png\\"
  title=\\"Count Dracula\\"
  color=\\"cyan\\"
/>;
",
  "title": "Colors",
}
`;

exports[`Site: Avatar Variants 1`] = `
Object {
  "docs": "Avatars accept multiple variants, which can be used to render a more subtle variation of the Avatar component.",
  "html": "<div>
  <span class=\\"drac-avatar drac-bg-green-transparent drac-text-green drac-mb-sm\\"
    ><span
      class=\\"drac-avatar-background\\"
      style=\\"
        background-image: url(https://ui.draculatheme.com/static/images/avatar.png);
      \\"
    ></span></span
  ><span
    class=\\"drac-avatar drac-bg-green-transparent drac-text-green drac-avatar-subtle\\"
    ><span
      class=\\"drac-avatar-background\\"
      style=\\"
        background-image: url(https://ui.draculatheme.com/static/images/avatar.png);
      \\"
    ></span
  ></span>
</div>
",
  "react": "<div>
  <Avatar
    src=\\"https://ui.draculatheme.com/static/images/avatar.png\\"
    title=\\"Count Dracula\\"
    variant=\\"normal\\"
    mb=\\"sm\\"
  />
  <Avatar
    src=\\"https://ui.draculatheme.com/static/images/avatar.png\\"
    title=\\"Count Dracula\\"
    variant=\\"subtle\\"
  />
</div>;
",
  "title": "Variants",
}
`;
